drivers: modem: operate on device pointers instead of names

mniestroj opened the pull request:
So far modem API used UART device names / labels. Change API to operate
on device pointers instead, so that we stop using device_get_binding()
in modem core and in some DT compatible modem drivers.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
drivers: modem: operate on device pointers instead of names

So far modem API used UART device names / labels. Change API to operate
on device pointers instead, so that we stop using device_get_binding()
in modem core and in some DT compatible modem drivers.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

gmarull commented on drivers/modem/modem_receiver.c:
```diff
@@ -224,9 +224,9 @@ int mdm_receiver_register(struct mdm_receiver_context *ctx,
 
-	ctx->uart_dev = device_get_binding(uart_dev_name);
+	ctx->uart_dev = uart_dev;
 	if (!ctx->uart_dev) {
-		LOG_ERR("Binding failure for uart: %s", uart_dev_name);
+		LOG_ERR("Binding failure for uart: %p", uart_dev);
 		return -ENODEV;
 	}
```
shouldn't that be replaced with a call to `device_is_ready()`? (returns false if not ready or pointer is NULL)

mniestroj replied:
Possibly. I will get back to it once unit tests will pass, as I am not aware when and how should `device_is_ready()` be used yet.

mniestroj replied:
I've added `device_is_ready()` here and in `modem_iface_uart_init_dev()`. Thanks for hint!

mniestroj committed:
tests: drivers: build_all: add fake serial device for modem tests

Serial device is needed for building drivers that use DEVICE_DT_GET().
None of the currently used modem drivers use that right now, but this is
about to change.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>
